Tile geometry projection support

peppsac opened the pull request:
This PR improves itowns handling of having multiple tile geometries with different CRS.

Until now, geometries are *assumed* to use the same CRS as the view which is often the case.
I believe we should support displaying other CRS too, as long as the transformation from 1 crs to the other is possible.
If not possible or not implemented we should fail early instead of silently accepting data that we can't display.

With that being said, this PR implements tile geometry projection and adds a smallish example to illustrate a possible use.
The example could be improved to look nicer and allow the user to type any CRS instead of hardcoding it: help is welcome, feel free to push improvements directly to the branch :-)

Here's the result with CRS 5514 (see https://epsg.io/5514):
![capture d ecran de 2017-10-06 17-26-36](https://user-images.githubusercontent.com/2198295/31285487-b8030522-aabb-11e7-9d34-377273a92ed7.png)

peppsac committed:
fix: convert tiles extent to view's crs

Stop assuming that geometry layers use the same crs as the view.
Instead, convert the given extent to the view crs.

peppsac committed:
fix (tiles): differentiate options.maxSubdivisionLevel = 0 and undefined

peppsac committed:
example: draw (projected) 3946 plane over the globe

gchoqueux commented:
there's a problem with picking mouse.
the planar layer `epsgLayer` must have the same THREE layer, for the `getPickingPositionFromDepth` to function

gchoqueux commented:
in example with globeView
`createPlanarLayer` return geometry layer which is not plan, because it is converted into the crs of the globe. On the other hand the layer is instantiated with a planar projection.
it could also be created with `createGlobeLayer`?
if yes it would be necessary to factorize to have a single function
WDYT?

peppsac commented:
> in example with globeView createPlanarLayer return geometry layer which is not plan, because it is converted into the crs of the globe. 

This geometry layer uses a planar CRS but it is displayed in a the globe's CRS so its underlying geometry uses `EPSG:4978`.
So IMHO, it makes sense to create it using `createPlanarLayer`.

gchoqueux commented:
`createPlanarLayer` causes tile visibility problems in the globe

gchoqueux commented:
is it possible to attach a imagery layer CRS 5514 to see the result?

peppsac commented:
Updated, with OSM data displayed on the projected layer.

![capture d ecran de 2017-10-20 14-42-22](https://user-images.githubusercontent.com/2198295/31821317-00575406-b5a5-11e7-943e-d35a0282acc9.png)

peppsac committed:
fix extent.offsetToParent

peppsac committed:
feat (gis): add support for ft-US unit

Some SIG (e.g 2263) uses ft-US unit so even if SI units CRS are recommended
this commit adds support for ft-US.

gchoqueux commented:
the projection is incorrect

peppsac commented:
> the projection is incorrect

Could you elaborate please?

From my perspective there's a visual artifact but maybe it's due to the 3857 CRS being used for the OpenStreepMap layer (using the fix from #478 ). Or are talking about something else?

gchoqueux committed:
fix (protocol) : fix the undefined layer.options.tileMatrixSet in WMS_Provider.tileTextureCount

gchoqueux commented:
I can't make the example work. the conversion projection are not compatible.
Could you check?

peppsac commented:
I'm not sure what's wrong here: #478 is supposed to fix 3857 for WMS layer and this PR (try to) use this feature.

gchoqueux commented:
how use PR's example?

peppsac committed:
rebase

peppsac commented:
> how use PR's example?

Open `epsg_finder.html`, click on 'View'.
The result should be similar to https://github.com/iTowns/itowns/pull/513#issuecomment-338195874

gchoqueux commented:
there are errors of the conversions of coordinates, when I try example

peppsac commented:
> there are errors of the conversions of coordinates, when I try example

That means you managed to reproduce the problem.
Now what I'd need is your input on how you think this could get fixed.

gchoqueux commented:
it's worse than that, I don't see the planar layer

peppsac commented:
So this PR depends on PR #561 to fix a `TileGeometry` bug.

zarov commented on src/Core/Prefab/Planar/PlanarTileBuilder.js:
```diff
@@ -40,17 +46,33 @@ PlanarTileBuilder.prototype.VertexNormal = function VertexNormal() {
 // coord u tile to projected
 PlanarTileBuilder.prototype.uProjecte = function uProjecte(u, params) {
-    params.projected.x = params.extent.west() + u * (params.extent.east() - params.extent.west());
+    params.projected._values[2] = 100;
```
Couldn't you use `Coordinates#setAltitude` ? I'm not fond of accessing properties that are supposed to be private. Same question for next lines.

zarov reviewed (commented):
Could you rebase your work ? Also any view on how to fix the bad coordinates conversion ?

peppsac commented:
I'm closing this PR for now - I'll reopen it when I'll have some time to fix the remaining issues.
